sdk/include/opentelemetry/sdk/trace/tracer_context.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/sdk/common/atomic_unique_ptr.h"
+#include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/sdk/trace/processor.h"
+#include "opentelemetry/sdk/trace/samplers/always_on.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+/**
+ * A class which stores the TracerProvider context.
+ *
+ * This class meets the following design criteria:
+ * - A shared reference between TracerProvider and Tracers instantiated.
+ * - A thread-safe class that allows updating/altering processor/exporter pipelines
+ *   and sampling config.
+ * - The owner/destroyer of Processors/Exporters.  These will remain active until
+ *   this class is destroyed.  I.e. Sampling, Exporting, flushing etc. are all ok if this
+ *   object is alive, and they willl work together.   If this object is destroyed, then
+ *   no shared references to Processor, Exporter, Recordable etc. should exist, and all
+ *   associated pipelines will have been flushed.
+ *
+ *
+ * TODOs - This should allow more than one processor pipeline to be attached.
+ */
+class TracerContext
+{
+public:
+  explicit TracerContext(std::shared_ptr<SpanProcessor> processor,
+                         opentelemetry::sdk::resource::Resource resource =
+                             opentelemetry::sdk::resource::Resource::Create({}),
+                         std::unique_ptr<Sampler> sampler =
+                             std::unique_ptr<AlwaysOnSampler>(new AlwaysOnSampler)) noexcept;
+  /**
+   * Obtain the span processor associated with this tracer context.
+   * <p>
+   * This does NOT give ownership to the caller and must be used within the lifecycle of
+   * a shared `TracerContext` object.
+   */
+  SpanProcessor *GetProcessor() const noexcept;
```

Comment:
Alternatively, we can also skip the `SetProcessor` method. The [spec ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation)doesn't mandate it:
> The TracerProvider MAY provide methods to update the configuration.

After some thinking, I didn't come across a use case where a `SetProcessor` method would be needed. Leaving it out would drastically simplify any concurrency considerations.

Reply:
TBH - We need to gut this to handle multiple processors (which most APIs/SDKs allow).   The current way we're using things fundamentally has to change, and in most cases where we have shared_ptr to Processor, it'd be better to have a Processor that has a shared_ptr to some external thing.

Haven't had time to write up my thoughts here, but I'm more than happy to remove the SetProcessor method.   I still think GetProcessor should NOT give ownership of a pointer (shared or otherwise) and only allow ephemeral usage of SpanProcessor.  I can (i think?) move this to SpanProcesor& if we great a null-object span processor.

Reply:
Could `GetProcess` return a `weak_ptr` instead of asking the caller to limit its usage to avoid UAF?

Reply:
Changed this to be a reference.  Going to add guarantees that there's ALWAYS a processor (even if it's a null-object-pattern processor that does nothing).